paddle/fluid/operators/quantize_mkldnn_op.cc

```diff
@@ -0,0 +1,100 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "mkldnn.hpp"
+#include "paddle/fluid/framework/tensor.h"
+#include "paddle/fluid/operators/quantize_op.h"
+#include "paddle/fluid/platform/mkldnn_helper.h"
+
+namespace paddle {
+namespace operators {
+
+using mkldnn::memory;
+using mkldnn::primitive;
+using mkldnn::reorder;
+using platform::to_void_cast;
+using Tensor = framework::Tensor;
+using framework::DataLayout;
+using mkldnn::stream;
+using platform::GetMKLDNNFormat;
+
+template <typename T>
+class QuantOpKernel : public framework::OpKernel<T> {
+ public:
+  void Compute(const framework::ExecutionContext& ctx) const override {
+    auto* input = ctx.Input<Tensor>("Input");
+    auto scale_data = ctx.Attr<float>("Scale");
+    auto* output = ctx.Output<Tensor>("Output");
+    auto& dev_ctx =
+        ctx.template device_context<platform::MKLDNNDeviceContext>();
+    const auto& engine = dev_ctx.GetEngine();
+
+    std::vector<primitive> pipeline;
+    std::vector<int> src_tz = paddle::framework::vectorize2int(input->dims());
+    std::vector<int> dst_tz = paddle::framework::vectorize2int(output->dims());
+
+    const T* input_data = input->data<T>();
+
+    mkldnn::primitive_attr attri;
+    int mask = 0;
+    attri.set_output_scales(mask, {scale_data});
+
+    auto src_md = platform::MKLDNNMemDesc({src_tz}, memory::data_type::f32,
+                                          input->format());
+    auto src_pd = mkldnn::memory::primitive_desc(src_md, engine);
+    auto src_memory =
+        std::make_shared<mkldnn::memory>(src_pd, to_void_cast<T>(input_data));
+    std::shared_ptr<primitive::at> src_memory_p =
+        std::shared_ptr<primitive::at>(new primitive::at(*src_memory));
+
+    bool is_negative = ctx.Attr<bool>("is_negative_input");
+    mkldnn::memory::primitive_desc dst_pd;
+    std::shared_ptr<mkldnn::memory> dst_memory;
+    if (is_negative) {
+      int8_t* output_data = output->mutable_data<int8_t>(ctx.GetPlace());
+      auto dst_md = platform::MKLDNNMemDesc({dst_tz}, memory::data_type::s8,
+                                            memory::format::nhwc);
+      dst_pd = mkldnn::memory::primitive_desc(dst_md, engine);
+      dst_memory.reset(
+          new mkldnn::memory(dst_pd, to_void_cast<int8_t>(output_data)));
+    } else {
+      uint8_t* output_data = output->mutable_data<uint8_t>(ctx.GetPlace());
+      auto dst_md = platform::MKLDNNMemDesc({dst_tz}, memory::data_type::u8,
+                                            memory::format::nhwc);
+      dst_pd = mkldnn::memory::primitive_desc(dst_md, engine);
+      dst_memory.reset(
+          new mkldnn::memory(dst_pd, to_void_cast<uint8_t>(output_data)));
```

Comment:
In the future it might be worth to extract both branches into a single templatized function that would accept int8/uint8 and memory::data_type as parameters.

Reply:
> extract both branches into a single templatized function

How about extract in this PR? if in the future, when will extract them?